File: LeetCodeRecord/MethodFirst/75-SortColors.java
package MethodFirst;

class SortColors {
    public void sortColors(int[] nums) {
        int i = 0;
        int j = 0;
        
        while (j < nums.length) {
            if (nums[j] == 0) {
                int temp = nums[i];
                nums[i] = nums[j];
                nums[j] = temp;

                i++;
            }

            j++;
        }

        j = i;
        while (j < nums.length) {
            if (nums[j] == 1) {
                int temp = nums[i];
                nums[i] = nums[j];
                nums[j] = temp;

                i++;
            }

            j++;
        }
    }
    
    public void sortColors2(int[] nums) {
        // [0, p1) 0
        // [p1, p2) 1
        int p1 = 0;
        int p2 = 0;
        int i = 0;

        while (i < nums.length) {
            if (nums[i] == 0) {
                int temp = nums[i];
                nums[i] = nums[p2];
                nums[p2] = nums[p1];
                nums[p1] = temp;

                p1++;
                p2++;
            } else if (nums[i] == 1) {
                int temp = nums[i];
                nums[i] = nums[p2];
                nums[p2] = temp;
                
                p2 ++;
            }

            i++;
        }
    }
}

File: LeetCodeRecord/ACFirst-Golang/75-SortColors.go
package main

func sortColors(nums []int) {
	numsLen := len(nums)

	if numsLen == 0 {
		return
	}

	p0 := 0
	p1 := 0
	i := 0

	for i < numsLen {
		if nums[i] == 0 {
			temp := nums[i]
			nums[i] = nums[p1]
			nums[p1] = nums[p0]
			nums[p0] = temp

			p0++
			p1++
		} else if nums[i] == 1 {
			temp := nums[i]
			nums[i] = nums[p1]
			nums[p1] = temp
			p1++
		}

		i++
	}
}

func sortColors2(nums []int) {
	numsLen := len(nums)

	if numsLen == 0 {
		return
	}

	i := 0
	j := 0

	for j < numsLen {
		if nums[j] == 0 {
			temp := nums[i]
			nums[i] = nums[j]
			nums[j] = temp
			i++
		}
		j++
	}

	j = i

	for j < numsLen {
		if nums[j] == 1 {
			temp := nums[i]
			nums[i] = nums[j]
			nums[j] = temp
			i++
		}

		j++
	}
}